routing_plugin.go

```diff
@@ -106,3 +123,522 @@ func ReleaseRoutingPlugin(ctx context.Context) {
+	exit := target
+
+	// Walk until the first fork (if there's any). This first
+	// fork will be where we're going to try to manipulate success
+	// probabilities to increasingly prefer more expensive edges.
+
+	// We track all visited peers, so we won't end up walking
+	// back and forth on graphs that have no forks.
+	visited := make(map[route.Vertex]struct{})
+
+	for {
+		targetNodeInfo, err := getNodeInfo(
+			ctx, r.lnd.Client, target, privateEdges,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Make sure we mark the current node as visited.
+		visited[target] = struct{}{}
+
+		// If the target node has only one or more channels but all
+		// connects to the same peer, then we need to walk further.
+		var (
+			peer  route.Vertex
+			peers []route.Vertex
+		)
+		for _, edge := range targetNodeInfo.Channels {
+			if target != edge.Node1 {
+				peer = edge.Node1
+			} else {
+				peer = edge.Node2
+			}
+
+			if _, ok := visited[peer]; !ok {
+				visited[peer] = struct{}{}
+				peers = append(peers, peer)
+			}
+		}
+
+		if len(peers) == 1 {
+			exit = target
+			target = peers[0]
+			continue
+		}
+
+		// If there are no more peers to visit then we can't use
+		// this routing plugin.
+		if len(peers) == 0 {
+			return ErrRoutingPluginNotApplicable
+		}
+
+		// Found the first fork to our target.
+		break
+	}
+
+	nodeInfo, err := getNodeInfo(ctx, r.lnd.Client, target, privateEdges)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("Routing plugin target: '%v' %v", nodeInfo.Alias,
+		nodeInfo.PubKey.String())
+
+	// Gather node info (including channels) for the nodes we're
+	// interested in.
+	targetChanged := exit != target
+	nodes := make(map[route.Vertex]*lndclient.NodeInfo)
+	for _, edge := range nodeInfo.Channels {
+		// Skip edges to the exit node since from there the route to
+		// the invoice target is always constructed from the same hops.
+		if targetChanged && (edge.Node1 == exit ||
+			edge.Node2 == exit) {
+			continue
+		}
+
+		var peer route.Vertex
+		if edge.Node1 == target {
+			peer = edge.Node2
+		} else {
+			peer = edge.Node1
+		}
+
+		nodeInfo, err := getNodeInfo(
+			ctx, r.lnd.Client, peer, privateEdges,
+		)
+		if err != nil {
+			return err
+		}
+
+		nodes[peer] = nodeInfo
+	}
+
+	// Get the nodes ordered by routing fee towards the target.
+	r.nodesByMaxFee = nodesByMaxFee(amt, target, nodes)
+	r.target = target
+	r.amount = amt
+
+	// Save MC state.
+	err = r.saveMissionControlState(ctx, nodes, target)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// BeforePayment will reconfigure the mission control on each payment attempt.
+func (r *lowToHighRoutingPlugin) BeforePayment(ctx context.Context,
+	currAttempt int, maxAttempts int) error {
+
+	queryRoutesReq := lndclient.QueryRoutesRequest{
+		Source:            &r.lnd.NodePubkey,
+		PubKey:            r.target,
+		AmtMsat:           lnwire.MilliSatoshi(r.amount * 1000),
+		FeeLimitMsat:      lnwire.MilliSatoshi(r.amount * 1000),
```

Comment:
Why are we setting the fee limit to the full amount? Don't swaps have their own restrictions?

Reply:
Yes, you're correct. It's just a debug query so I felt it's not super important to be fully pedantic here, nor is it correct in the sense that we use the full amount as normally we may split too. I can remove these debug logs too, I just found them useful for understanding what's happening.